2018/2018_File_Generation_Features.md

```diff
@@ -0,0 +1,67 @@
+
+-[`clean_file_config.yaml`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/configurations/clean_file_config.yaml) holds values for the test institution, file length, and file name for generating a clean HMDA submission file for testing. Other variables, such as data ranges can also be set in this file. 
+
+Clean files will pass syntax and validity edits. Edit test files will fail at least one (and sometimes more than one) edit. These files will be used as a double check for the implementation of the HMDA edits for the 2018 HMDA Platform. Creating edit test files is a two step process. The generation of edit test files requires a clean data file to be present. If edit files already exist with the same names as those produced by this script, they will be overwritten.
+
+Edit test files are written to the directory indicated in `test_filepaths.yaml` based on edit type, such as /edits_files/test_files/Bank1/syntax/s300.txt. The file name indicates which edit the file is designed to test. If an edit has multiple conditions, a file will be made for each condition in the format /edits_files/test_files/Bank1/syntax/s301_1.txt.
+
+### Changing Edit Logic and the Classes Used for File Creation 
+The FileGenerator Class relies on four other classes, each with its own role in file creation. 
+
+-`lar_generator` in [`lar_generator.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/lar_generator.py) creates raw TS and LAR rows, the values of which are then manipulated to create clean test files or edit test files. 
+
+-`rules_engine` in [`rules_engine.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/rules_engine.py) contains logic for each edit. LAR and TS dataframes are passed into this class, and its methods output edit details that show among its fields the edit name, the status of the edit (failed or passed), the number of rows failed by the edit, if any, and the ULI's or 
+NULIs (loan ID) of the rows that fail the edit. Each method creates a results dataframe that holds the rows that have failed the specific edit. The class then wraps the results into an edit report that can be parsed by other methods in the repository.  
+
+- `lar_constraints` in [`lar_constraints.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/rules_engine.py) contains methods for creating clean LAR rows that conform to pass syntax or validity edits. New edit logic methods must be made in terms of changing a TS or LAR row. 
+
+- `test_data` in [`test_file_generator.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/test_file_generator.py) contains methods that manipulate LAR or TS dataframes to fail a specific syntax or validity edit. New edit logic methods must be made in terms of changing a TS or LAR dataframe.
```

Comment:
Does `New edit logic methods must be made in terms of changing a TS or LAR dataframe.` mean new constraints?

Reply:
Can you make a note that the changes to the clean data do not create exhaustive test cases for each edit?

Reply:
Sure - I also removed the line you mentioned in the first comment. I think that this is pretty clear from the other things said.